test/model/recent_senders_test.dart

```diff
@@ -0,0 +1,153 @@
+    }
+
+    test('starts with no stream or topic senders', () {
+      prepare();
+      check(recentSenders.debugIsEmpty).equals(true);
+    });
+
+    test('only processes a stream message', () {
+      prepare();
+      final dmMessage = eg.dmMessage(from: eg.selfUser, to: [eg.otherUser]);
+      recentSenders.handleMessage(dmMessage);
+      check(recentSenders.debugIsEmpty).equals(true);
+
+      final streamMessage = eg.streamMessage(
+        id: 100,
+        sender: eg.user(userId: 10),
+        stream: eg.stream(streamId: 1),
+        topic: 'topic',
+      );
+      recentSenders.handleMessage(streamMessage);
+
+      final expectedStreamSenders = {
+        1: {
+          10: MessageIdTracker.fromIds([100])
+        }
+      };
+      check(recentSenders.debugStreamSenders).deepEquals(expectedStreamSenders);
+
+      final expectedTopicSenders = {
+        1: {
+          'topic': {
+            10: MessageIdTracker.fromIds([100])
+          },
+        },
+      };
+      check(recentSenders.debugTopicSenders).deepEquals(expectedTopicSenders);
+    });
+
+    test('adding multiple messages', () {
+      prepare();
+      final message1 = eg.streamMessage(
+        stream: eg.stream(streamId: 1),
+        topic: 'topic1',
+        sender: eg.user(userId: 10),
+        id: 300,
+      );
+      final message2 = eg.streamMessage(
+        stream: eg.stream(streamId: 1),
+        topic: 'topic2',
+        sender: eg.user(userId: 10),
+        id: 100,
+      );
+      final message3 = eg.streamMessage(
+        stream: eg.stream(streamId: 1),
+        topic: 'topic1',
+        sender: eg.user(userId: 20),
+        id: 200,
+      );
+      final message4 = eg.streamMessage(
+        stream: eg.stream(streamId: 1),
+        topic: 'topic2',
+        sender: eg.user(userId: 20),
+        id: 400,
+      );
+      final message5 = eg.streamMessage(
+        id: 500,
+        stream: eg.stream(streamId: 2),
+        topic: 'topic3',
+        sender: eg.user(userId: 20),
+      );
+
+      recentSenders.handleMessage(message1);
+      recentSenders.handleMessage(message2);
+      recentSenders.handleMessage(message3);
+      recentSenders.handleMessage(message4);
+      recentSenders.handleMessage(message5);
+
+      final expectedStreamSenders = {
+        1: {
+          10: MessageIdTracker.fromIds([100, 300]),
+          20: MessageIdTracker.fromIds([200, 400]),
+        },
+        2: {
+          20: MessageIdTracker.fromIds([500]),
+        },
+      };
+
+      final expectedTopicSenders = {
+        1: {
+          'topic1': {
+            10: MessageIdTracker.fromIds([300]),
+            20: MessageIdTracker.fromIds([200]),
+          },
+          'topic2': {
+            10: MessageIdTracker.fromIds([100]),
+            20: MessageIdTracker.fromIds([400]),
+          },
+        },
+        2: {
+          'topic3': {
+            20: MessageIdTracker.fromIds([500]),
+          }
+        }
+      };
```

Comment:
This is kind of a lot of data to read, even though this is one test that isn't super complicated.

We'll want more tests than this — especially once we add support for adding messages in bulk rather than one by one (the algorithm discussed at https://github.com/zulip/zulip-flutter/pull/692#discussion_r1662930537), and once we add support for removing messages, both of which will make this data structure's logic more complex. To do that, it'll be crucial to make each test case less of a burden to write.

Fortunately, all this expected data is also totally mechanically determined by the input data above. In this version, you as the test author have to write this all down, and then a reviewer has to compare it item by item against the input data in order to confirm it's correct (so that it's checking the implementation is right, instead of checking it's wrong). But instead you can have the computer do that work.

For an example, see `model/unreads_test.dart`, and in particular the `checkMatchesMessages` helper.

Following that example, both of these expected-data variables, and the `check` calls below, can look instead like:
```dart
  checkMatchesMessages([message1, message2, message3, message4, message5]);
```

Or given the other comment above about how `prepare` and the late variable aren't necessary here, it'd be:
```dart
  checkMatchesMessages(model, [message1, message2, message3, message4, message5]);
```
with a `final model = RecentSenders();` at the top of the test case.